java/src/jmri/util/AlphanumComparator.java

```diff
@@ -101,6 +101,9 @@ public int compare(String s1, String s2) {
                 break;
             }
         }
+        if (result == 0 && marker1 == length1 && marker2 < length2) return -1;
+        if (result == 0 && marker1 < length1 && marker2 == length2) return +1;
+        
         return result;
```

Comment:
Since `Comparible.compareTo(...)` allows any non-zero value to be returned, but `Comparator.compare(o1, o2)` is only -1, 0, 1, this return statements should be something like:
`return result > 0 ? 1 : result < 0 ? -1 : 0;`

Sorry about commenting first on the test, and the resultant noise.

Reply:
Thanks for looking at all this. More to follow on the entire PR, but I'm offline for a bit today.  

What do you think of using the [Integer.signum(int)](https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#signum(int)) operator for limiting the range of the result? It seems to be a common metaphor, but I'm not sure whether it's clearer or not, not sure of the performance issues.

Reply:
The six redundant conditionals and inline returns really tweak my nerves.
The whole thing could be simplified/reduced to:
if (result == 0) { result = Integer.signum(length1 - length2);}

Reply:
The redundant conditionals are a matter of style, I guess.  With the code as it is, I think it's true that you can never reach the end without either one or both of strings at their end i.e. marker==length.  But that's a subtle invariant, and easy to break in the future. So I tend to check things like that.

For example, if the code is ever modified to suppress redundant leading zeros (which would be good), the 'result = Integer.signum(length1 - length2)' would become a subtle error. Since marker1 and marker2 don't have to advance together, I'm not even 100% sure how to prove it's correct now.